File: CalendarWithObjC/CalendarWithObjC/CustomCalendar/Calendar/CalendarProvider/CalendarProvider.m
//
//  CalendarProvider.m
//  CathayBank
//
//  Created by Marvin iOS on 2018/9/12.
//  Copyright © 2018年 YuChen Hsu. All rights reserved.
//

#import "CalendarProvider.h"

@implementation CalendarProvider


- (CalendarMonthModel*)getCalendarMonthModel:(NSDate*)date
                                      offset:(NSInteger)offset
{
    NSDateComponents *dateComponents = [[NSCalendar currentCalendar] components: NSCalendarUnitYear | NSCalendarUnitMonth | NSCalendarUnitDay fromDate:date];
    
    NSDateComponents *inputDateComponents;
    
    if (offset == 0)
    {
        inputDateComponents = dateComponents;
    }
    else
    {
        NSDate *inputDate = [[NSCalendar currentCalendar] dateByAddingUnit:NSCalendarUnitMonth value:offset toDate:date options:0];
        
        inputDateComponents = [[NSCalendar currentCalendar] components:NSCalendarUnitYear | NSCalendarUnitMonth | NSCalendarUnitDay fromDate:inputDate];
    }
    
    CalendarMonthModel *model = [CalendarMonthModel new];
    
    [self setMonthModal:model dateComponent:inputDateComponents];
    
    return model;
}

-(void)setMonthModal:(CalendarMonthModel*)monthModel
       dateComponent:(NSDateComponents*)dateComponent
{
    monthModel.dayRange = [self getDayRangeInThisMonth:dateComponent];
    monthModel.weekDayOfFirst = [self getWeekdayOfMonthFirst:dateComponent];
    monthModel.emptyCellCount = [self getEmptyCellCount:dateComponent];
    monthModel.totalItemCount = monthModel.emptyCellCount + monthModel.dayRange;
    // month label 放在 section 內
//    monthModel.monthLabelIndexPath = [self getMonthLabelIndex:dateComponent];
    monthModel.year = [dateComponent year];
    monthModel.month = [dateComponent month];
    
    // Array index 從 0 開始， month number 要減 1
    monthModel.monthLabelString = [[[NSDateFormatter new] monthSymbols] objectAtIndex:[dateComponent month] - 1];
}

- (int)getDayRangeInThisMonth:(NSDateComponents*)dateComponents
{
    
    [dateComponents setDay:1];
    NSDate *date = [[NSCalendar currentCalendar] dateFromComponents:dateComponents];
    NSRange range = [[NSCalendar currentCalendar] rangeOfUnit:NSCalendarUnitDay inUnit:NSCalendarUnitMonth forDate:date];
    
    return (int)range.length;
}

- (int)getEmptyCellCount:(NSDateComponents*)dateComponents
{
    int weekDayOfFirst = [self getWeekdayOfMonthFirst:dateComponents];
    return weekDayOfFirst - 1;
}


// month label 放在 section 內
//- (int)getMonthLabelIndex:(NSDateComponents*)dateComponents
//{
//    int weekDayOfFirst = [self getWeekdayOfMonthFirst:dateComponents];
//    return weekDayOfFirst - 1;
//}

- (int)getWeekdayOfMonthFirst:(NSDateComponents*)dateComponents
{
    NSCalendar *calendar = [NSCalendar currentCalendar];
    NSDate *date = [calendar dateFromComponents:dateComponents];
    NSInteger weekDay = [calendar component:NSCalendarUnitWeekday
                                   fromDate:date];
    // Obj-c 裡面, Sunday 設定是 1, Staturday 是 7
    return (int)weekDay;
}

@end

File: CalendarWithSwift/CalendarWithSwift/CustomCalendar/Calendar/CalendarProvider/CalendarProvider.swift
//
//  CalendarProvider.swift
//  CalendarWithSwift
//
//  Created by YuChen Hsu on 2019/7/25.
//  Copyright © 2019 YuChen Hsu. All rights reserved.
//

import Foundation

struct CalendarProvider {
//    let selectedDate1 : DateComponents
//    let selectedDate2 : DateComponents
    
    func getCalendarMonthModel(date:Date, offset:NSInteger) -> CalendarMonthModel {
        let dateComponents = Calendar.current.dateComponents([Calendar.Component.year,Calendar.Component.month,Calendar.Component.day], from: date)
        
        var inputDateComponents : DateComponents
        
        if offset == 0 {
            inputDateComponents = dateComponents
        }else{
            let inputDate:Date = Calendar.current.date(byAdding: Calendar.Component.month, value: offset, to: date)!
            
            inputDateComponents = Calendar.current.dateComponents([Calendar.Component.year,Calendar.Component.month,Calendar.Component.day], from: inputDate)
        }
        
        var model = CalendarMonthModel.init(weekDayOfFirst: 0, dayRange: 0, emptyCellCount: 0, totalItemCount: 0, monthLabelIndexPath: 0, monthLabelString: "", year: 0, month: 0)
        
        self.setMonthModal(monthModel: &model, dateComponent: inputDateComponents)
        
        return model
        
        
    }
    
    func setMonthModal(monthModel:inout CalendarMonthModel,dateComponent:DateComponents){
        var inputDateComponents = dateComponent
        
        monthModel.dayRange = self.getDayRangeInThisMonth(dateComponents:&inputDateComponents)
        monthModel.weekDayOfFirst = self.getWeekdayOfMonthFirst(dateComponents: inputDateComponents)
        monthModel.emptyCellCount = self.getEmptyCellCount(dateComponents: inputDateComponents)
        monthModel.totalItemCount = monthModel.emptyCellCount + monthModel.dayRange
//        monthModel.monthLabelIndexPath = self.getMonthLabelIndex(dateComponents: dateComponent)
        monthModel.year = inputDateComponents.year!
        monthModel.month = inputDateComponents.month!
        // Array index 從 0 開始， month number 要減 1
        monthModel.monthLabelString = DateFormatter.init().monthSymbols[inputDateComponents.month!-1]
    }
    
    func getDayRangeInThisMonth (dateComponents: inout DateComponents) -> Int!{
        
        dateComponents.day = 1
        let date = Calendar.current.date(from: dateComponents)
        
        let range = Calendar.current.range(of: Calendar.Component.day, in: Calendar.Component.month, for: date!)
        return range?.count
    }

    func getWeekdayOfMonthFirst (dateComponents:DateComponents) -> Int!{
        
        let date = Calendar.current.date(from: dateComponents)
        
        let weekDay = Calendar.current.component(Calendar.Component.weekday, from: date!)
        
        return weekDay
    }
    
    // month label 放在 section 內
    func getMonthLabelIndex (dateComponents:DateComponents) -> Int!{

        let weekDayOfFirst = self.getWeekdayOfMonthFirst(dateComponents: dateComponents) - 1

        return weekDayOfFirst
    }
    
    func getEmptyCellCount (dateComponents:DateComponents) -> Int!{
        
        let weekDayOfFirst:Int = self.getWeekdayOfMonthFirst(dateComponents: dateComponents)
        
        return weekDayOfFirst - 1
    }
}
